### DISCUSSION TITLE
Point cloud layer performance optimization

### DISCUSSION THREAD

**Discussion by radoslavirha** [visgl/deck.gl#7454] [Q&A]
Hi,

Do you have some suggestions for point clouds optimization in FPV? I have powerful computer and experience is not so good.

Some point clouds are very laggy, I know it depends on data (denser Point cloud = worse performance it seems) is, but changing `maximumScreenSpaceError` does not improve performance.

I tried few examples from loaders.gl repo and e.g. 555 Market is very laggy.

https://codesandbox.io/s/deck-gl-3d-tiles-forked-d4f994?file=/src/App.js

Try 555 Market here: https://loaders.gl/examples/3d-tiles. It's also laggy in MapView, so I think it's not just FirstPersonView issue.

We can improve creating/tiling point clouds probably, but from deck.gl/loaders.gl side, is there something what can automatically optimize rendering?

Thanks

**Comment by ibgreen** [comment]
Some thoughts:
- **Computer**: 555 Market is buttery smooth on my latest generation MacBook Pro Max, but it is generally an order of magnitude faster than older MacBooks, and even more so compared to phones. Also make sure you are not using Radeon 500 series GPU MacBooks as there is a known severe perf issue with those drivers on later MacOS versions.
- **Fragments** - One thing to be wary of is the number of fragments (i.e. pixels) rendered per point - e.g. rendering 10M points as a discs with 6 pixel radius can generate > 1 billion pixels (6*6*PI * 10M) per frame which will likely overwhelm the GPU.  The point cloud layer has properties to control this, you could probably experiment with Tile3D layer _subLayerProps to see if this is the issue. Also reducing devicePixelRatio helps by reducing total number of pixels with factor of 4, if this helps it would give a quick indication that you are fragment bound.
- **Tiling** - Some tilesets are not "correctly" tiled and load "too many" tiles at a given angle. I suspect 555 Market was tiled with the experimental (unfortunately, unmaintained) C++ point tiler in loaders.gl repo, so it may well be quite unoptimal and load too much. Tiling your data with a professional toolchain, perhaps Cesium ION, will likely allow you to fine tune these things.
- **Tileset traversal** - The Tileset traversal code in loaders is complex and I suspect that not too much effort has been spent testing the maxScreenSpaceError type properties. It is possible that there could be a bug or missed functionaly here, but fixing it would require someone willing to dig deep into that area.

**Comment by radoslavirha** [comment]
@ibgreen I'm on MacBook Air M1 with highest configuration, so I don't expect lags :D some point clouds are smooth, but few of them are laggy...depends on data it seems.

If you open my codesandbox and move between the buildings (more points), don't you experience some lags compared to locations on the edges of the point cloud?

Currently we use ION for tiling, previously we were experimenting with other solutions, but nothing was comparable with ION

**Comment by ibgreen** [comment]
Nope, 555 Market is extremely smooth on my machine both in the loaders.gl example page (via dropdown) and in your code sandbox. 

@dryabinin-actionengine @belom88 Can you do some testing? I think my machine is too powerful to exhibit the issue.

**Comment by Pessimistress** [comment]
If you don't care much about the points being round, the absolutely fastest way to render a point cloud is using `GL.POINTS`. `PointCloudLayer` draws a triangle, so it renders 3x the vertex count.

**Reply by radoslavirha** [reply]
@Pessimistress how can I set `GL.POINTS`? Thanks

**Reply by Pessimistress** [reply]
You will need to create a custom layer.

Here is a minimal example: https://codepen.io/Pessimistress/pen/XWYyWrB?editors=0010
